ckanext/data_qld/controller.py

```diff
@@ -74,3 +79,49 @@ def show_schema(self, dataset_id, resource_id):
         except tk.NotAuthorized as e:
             log.warn(e)
             tk.abort(403, tk._('You are not authorized to view the Data Scheme for the resource %s' % resource_id))
+
+
+class DataQldDataset(PackageController):
+    def _get_context(self):
+        return {'model': model, 'session': model.Session, 'user': c.user, 'for_view': True, 'auth_user_obj': c.userobj}
+
+    def _is_dataset_private(self, id):
+        try:
+            get_action('package_show')(self._get_context(), {'id': id})
+            return True
+        except NotFound:
+            abort(404, _('Dataset not found'))
+        except NotAuthorized:
+            return False
+
+    def read(self, id):
+        """
+        Override the default CKAN behaviour for private Dataset visibility for unauthenticated users.
+        Instead of displaying "404 Dataset not found" message, give unauthenticated users a chance to login (if the
+        dataset exists but is private).
+        :param id: Package id/name
+        :return:
+        """
+        if not g.user and not self._is_dataset_private(id):
+            tk.redirect_to(
+                tk.url_for('user.login', came_from='/dataset/{id}'.format(id=id))
+            )
+
+        return super(DataQldDataset, self).read(id)
+
+    def resource_read(self, id, resource_id):
+        """
+        Override the default CKAN behaviour for private Dataset Resource visibility for unauthenticated users.
+        Instead of displaying "404 Dataset not found" message, give unauthenticated users a chance to login (if the
+        dataset exists but is private).
+        :param id: Package id/name
+        :param resource_id: Resource id
+        :return:
+        """
+        if not g.user and not self._is_dataset_private(id):
```

Comment:
This works, but the name of the function is inverted. It should redirect when the target dataset _is_ private.